language/move-analyzer/src/diagnostics.rs

```diff
@@ -23,11 +24,36 @@ pub fn lsp_diagnostics(
     file_name_mapping: &BTreeMap<FileHash, Symbol>,
 ) -> BTreeMap<Symbol, Vec<Diagnostic>> {
     let mut lsp_diagnostics = BTreeMap::new();
-    for (s, _, (loc, msg), _, _) in diagnostics {
+    for (s, _, (loc, msg), labels, _) in diagnostics {
         let fpath = file_name_mapping.get(&loc.file_hash()).unwrap();
         if let Some(start) = get_loc(&loc.file_hash(), loc.start(), files, file_id_mapping) {
             if let Some(end) = get_loc(&loc.file_hash(), loc.end(), files, file_id_mapping) {
                 let range = Range::new(start, end);
+                let related_info_opt = if labels.is_empty() {
+                    None
+                } else {
+                    let mut related_info = vec![];
+                    for (lloc, lmsg) in labels {
+                        if let Some(lstart) =
+                            get_loc(&lloc.file_hash(), lloc.start(), files, file_id_mapping)
+                        {
+                            if let Some(lend) =
+                                get_loc(&lloc.file_hash(), lloc.end(), files, file_id_mapping)
+                            {
```

Comment:
A bit tricky to make "cleaner", but I think using `?` here would help a lot 
```
Some(labels.into_iter()
    .filter_map((lloc, lmsg) {
        let lstart = get_loc(&lloc.file_hash(), lloc.start(), files, file_id_mapping)?;
        let lend = get_loc(&lloc.file_hash(), lloc.end(), files, file_id_mapping)?;
        ...
        Some(DiagnosticRelatedInformation { ... })
     })
     .collect())
```

Reply:
Nice trick with `?` and `filtered_map`!